Implement registration page

marhkb opened the pull request:
This commit just adds simple page for registration.
What is still missing is an option to display the Telegram terms of
Service.
But the design how to display the ToS needs to be discussed first.

![register](https://user-images.githubusercontent.com/3630213/134645180-bc679d70-74e8-4d34-aaf4-3e57f1649f3c.png)

melix99 commented on data/resources/ui/login.ui:
```diff
@@ -195,6 +195,79 @@
+                              <object class="GtkLabel" id="registration_error_label">
+                                <property name="visible">False</property>
+                                <style>
+                                  <class name="error" />
```
```suggestion
                                  <class name="error"/>
```

melix99 commented on data/resources/ui/login.ui:
```diff
@@ -195,6 +195,79 @@
+                                  </object>
+                                </child>
+                                <style>
+                                  <class name="content" />
```
```suggestion
                                  <class name="content"/>
```

melix99 commented on src/login.rs:
```diff
@@ -441,3 +472,11 @@ impl Login {
     }
 }
+
+// HACK: It seems gtk4 removed get_toplevel
```
There's [get_root](https://docs.gtk.org/gtk4/method.Widget.get_root.html) now.

marhkb committed:
Implement registration page

This commit just adds a simple page for user registration.

It also brings a Terms of Service (ToS) dialog, that the user can open
via a clickable label. If Telegram enforces the user to accept the ToS,
the login flow will not proceed until the user has done this. Then,
the dalog is forced upon the user when she/he wants to proceed.

What is still missing is a photo upload function, as found in the
official client.

marhkb committed:
Update German translation for registration page

melix99 commented on src/login.rs:
```diff
@@ -182,20 +209,68 @@ impl Login {
-        } else if visible_page == "encryption-key-page" {
-            self.send_encryption_key(false);
+        let builder = gtk::MessageDialog::builder()
+            .text(&*self_.tos_text.borrow())
```
```suggestion
            .secondary_text(&*self_.tos_text.borrow())
```

The `text` is the header of the dialog and the `secondary-text` is the text below the header. You used the `text` as the description and the `title` (GtkWindow's parameter) as the header which gives the dialog the weird "compact" look.

melix99 commented on src/login.rs:
```diff
@@ -182,20 +209,69 @@ impl Login {
+        let dialog = if self_.show_tos_popup.get() {
+            builder
+                .buttons(gtk::ButtonsType::YesNo)
+                .title(&gettext("Do You Accept the Terms of Service?"))
```
```suggestion
                .text(&gettext("Do You Accept the Terms of Service?"))
```

melix99 commented on src/login.rs:
```diff
@@ -182,20 +209,69 @@ impl Login {
+        } else {
+            builder
+                .buttons(gtk::ButtonsType::Ok)
+                .title(&gettext("Terms of Service"))
```
```suggestion
                .text(&gettext("Terms of Service"))
```

melix99 commented on data/resources/ui/login.ui:
```diff
@@ -193,6 +193,101 @@
+                        <property name="tightening-threshold">200</property>
+                        <property name="child">
+                          <object class="GtkLabel" id="tos_label">
+                            <property name="ellipsize">PANGO_ELLIPSIZE_MIDDLE</property>
```
```suggestion
                            <property name="ellipsize">middle</property>
```

melix99 commented on src/login.rs:
```diff
@@ -42,6 +43,17 @@ mod imp {
+        pub registration_error_label: TemplateChild<gtk::Label>,
+        #[template_child]
+        pub tos_label: TemplateChild<gtk::Label>,
+        pub tos_text: RefCell<String>,
```
I tend to put non-widget parameters on top to keep them separate from widgets to improve readability, so let's do that here too.

melix99 commented on src/login.rs:
```diff
@@ -100,6 +115,11 @@ mod imp {
                 .bind("use-test-dc", use_test_dc_switch, "state")
                 .build();
+
+            self_.tos_label.connect_activate_link(|this, _| {
```
```suggestion
            self_.tos_label.connect_activate_link(|label, _| {
```

Seeing `this` makes it really difficult to understand what it actually is.

melix99 commented on src/login.rs:
```diff
@@ -182,20 +209,69 @@ impl Login {
         }
+        .build();
+
+        dialog.run_async(clone!(@weak self as obj => move |this, response| {
```
```suggestion
        dialog.run_async(clone!(@weak self as obj => move |dialog, response| {
```

melix99 commented on src/login.rs:
```diff
@@ -158,8 +181,12 @@ impl Login {
-                todo!()
+            AuthorizationState::WaitRegistration(data) => {
+                self_.show_tos_popup.set(data.terms_of_service.show_popup);
+                self_.tos_text.replace(data.terms_of_service.text.text);
```
The returned TOS here is a formatted text, which means that it can contain text that needs to be parsed to be displayed with the correct format. Now, I've tested to see if there's actually some entries to parse and it doesn't seem to be the case, but I'd still prefer to add a conversion just to be sure that future updates of the TOS will be displayed correctly if they'll include formatted text.

I've already implemented a parser for the FormattedText in the [message_bubble.rs](https://github.com/melix99/telegrand/blob/a22e38928da6b7843706841bbfa778a678e4d0e2/src/session/content/message_bubble.rs#L173) so I'd day to move that functions to utils.rs and use them to convert the text to markup, this way the eventual formatted text should be displayed correctly in the dialog.

melix99 commented on src/login.rs:
```diff
@@ -182,20 +209,69 @@ impl Login {
             }
+            "password-page" => self.send_password(),
+            "encryption-key-page" => self.send_encryption_key(false),
+            other => unreachable!("no page called '{}'", other),
```
```suggestion
            other => unreachable!("No page named '{}'", other),
```

marhkb committed:
Move function 'parse_formatted_text' to utils.rs

It is planned to use this function in other places like for parsing
the ToS text. Consequently, its companion function
'convert_to_markup' is also moved to utils.rs.

melix99 commented on src/login.rs:
```diff
@@ -182,20 +211,69 @@ impl Login {
+        }
+    }
+
+    fn show_tos_dialog(&self, user_wants_to_proceed: bool) {
```
```suggestion
    fn show_tos_dialog(&self, user_needs_to_accept: bool) {
```

melix99 commented on src/login.rs:
```diff
@@ -182,20 +211,69 @@ impl Login {
+
+            if matches!(response, gtk::ResponseType::No) {
+                self_.tos_accepted_by_user.set(false);
+                if user_wants_to_proceed {
```
This should not be needed anymore if you apply the change of the review above.

melix99 replied:
And ofc github mixed up the reviews :D

I was referring to the review of line 250.

marhkb replied:
At least the `obj.unfreeze();` is needed here, otherwise the UI stays frozen after the user declines the ToS.

marhkb replied:
@melix99 sorry, github loaded your second comment after I commented. I am a bit confused now. :smiley: 
Otherwise, everything is fine with this if clause as it is now?

melix99 replied:
I was referring to the fact that the if was not needed because both the Yes and No response can be received only when `user_needs_to_accept` is true, but I noticed that you left the check in the Yes response, maybe because for some reason the Yes response gets activated even when pressing the OK button?

marhkb replied:
Oh, I see that I have overseen a ` && user_needs_to_accept` that I have removed now. I also left some comments. Feel free to reopen if something else is missing.

melix99 commented on src/login.rs:
```diff
@@ -182,20 +211,69 @@ impl Login {
+            }
+            "code-page" => self.send_code(),
+            "registration-page" => {
+                if !self_.show_tos_popup.get() || self_.tos_accepted_by_user.get() {
```
I guess you added the `tos_accepted_by_user` variable to avoid asking again to the user if he/she wants to accept the tos when the registration failed previously. I'd say to just be safe and remove the `tos_accepted_by_user` variable completely and ask again if the user wants to accept the tos. The only conceivable way to fail the registration is not setting a name, so it's not likely to happen frequently anyway.

marhkb replied:
You are correct.

melix99 commented on src/login.rs:
```diff
@@ -182,20 +211,69 @@ impl Login {
+            .modal(true)
+            .transient_for(self.root().unwrap().downcast_ref::<gtk::Window>().unwrap());
+
+        let dialog = if self_.show_tos_popup.get() {
```
```suggestion
        let dialog = if user_needs_to_accept {
```
Otherwise the user would get the Yes/No dialog even when he/she wants to only see the tos pressing the label at the bottom.

marhkb replied:
Yeah, my intention was that the user could accept the ToS beforehand. So that she/he will not be bothered by the ToS dialog again. :smiley:

melix99 commented on src/login.rs:
```diff
@@ -182,20 +211,69 @@ impl Login {
+                }
+            } else if matches!(response, gtk::ResponseType::Yes) {
+                self_.tos_accepted_by_user.set(true);
+                if user_wants_to_proceed {
```
Same as above.

melix99 replied:
Was referring to the "above" review of line 266.

melix99 reviewed (changes requested):
I just noticed a few more things and then I think the PR is ready to merge.
